src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,126 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_NUM = 1.5;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[])new Object[DEFAULT_CAPACITY];
+
+    }
+
     @Override
     public void add(T value) {
+        if (size == array.length) {
+            array = grow();
+        }
+        array[size++] = value;
 
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size == array.length) {
+            array = grow();
+        }
+        if (index - 1 > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Object index cannot be "
+                    + "larger than the list size");
+        }
+        System.arraycopy(array, index, array, index + 1, size - index);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBound(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkBound(index);
+        array[index] = value;
 
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBound(index);
+        T removeElement = array[index];
+        int startIndex = index;
+        int elementsToMove = size - index - 1;
+        System.arraycopy(array, startIndex + 1, array, startIndex, elementsToMove);
+        size--;
+        return removeElement;
     }
 
     @Override
     public T remove(T element) {
```

Comment:
i believe you can implement it a more concise way :) looks like code duplication
remove external if-else for starters
reuse the existing method with this exact logic
![image](https://github.com/mate-academy/jv-array-list/assets/49767049/1807c098-7467-4ae8-aff6-f7204f4e2097)